core/store/src/opener.rs

```diff
@@ -173,72 +197,120 @@ impl<'a> StoreOpener<'a> {
     /// other hand, if mode is [`Mode::Create`], fails if the database already
     /// exists.
     pub fn open_in_mode(&self, mode: Mode) -> Result<crate::NodeStorage, StoreOpenerError> {
-        if let Some(db_version) = self.db.get_version()? {
-            let mode = mode.but_cannot_create().ok_or(StoreOpenerError::DbAlreadyExists)?;
-            let snapshot = self.apply_migrations(mode, db_version)?;
-            tracing::info!(target: "near", path=%self.path().display(),
-                           "Opening an existing RocksDB database");
-            let storage = self.open_storage(mode, Some(DB_VERSION))?;
-            snapshot.remove().map(|_| storage).map_err(StoreOpenerError::from)
+        if let Some(metadata) = self.db.get_metadata()? {
+            self.open_existing(
+                mode.but_cannot_create().ok_or(StoreOpenerError::DbAlreadyExists)?,
+                metadata,
+            )
+        } else if mode.can_create() {
+            self.open_new()
         } else {
-            let mode = mode.but_must_create().ok_or(StoreOpenerError::DbDoesNotExist)?;
+            Err(StoreOpenerError::DbDoesNotExist)
+        }
+    }
+
+    fn open_existing(
+        &self,
+        mode: Mode,
+        metadata: DbMetadata,
+    ) -> Result<crate::NodeStorage, StoreOpenerError> {
+        let snapshot = self.apply_migrations(mode, metadata)?;
+        tracing::info!(target: "near", path=%self.path().display(),
+                       "Opening an existing RocksDB database");
+        let (storage, metadata) = self.open_storage(mode, DB_VERSION)?;
+
+        // Update kind if the database is an RPC but we expect an Archive.
+        if self.expected_kind.is_some() && self.expected_kind != metadata.kind {
+            assert_eq!(
```

Comment:
I find two separate equality assertions easier to read.
`Archive == expected_kind`
`RPC == metadata.kind`